File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSFlashImageMessageCell.m
//
//  MSFlashImageMessageCell.m
//  BlackFireIM
//
//  Created by benny wang on 2022/1/25.
//

#import "MSFlashImageMessageCell.h"
#import "MSIMSDK-UIKit.h"


@implementation MSFlashImageMessageCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        _maskView = [[UIImageView alloc] init];
        _maskView.layer.cornerRadius = 5.0;
        [_maskView.layer setMasksToBounds:YES];
        _maskView.contentMode = UIViewContentModeScaleAspectFill;
        [self.container addSubview:_maskView];
        _maskView.frame = self.container.bounds;
        _maskView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;

        _fireIcon = [[UIImageView alloc]init];
        [self.container addSubview:_fireIcon];
        
        _progressL = [[UILabel alloc] init];
        _progressL.textColor = [UIColor whiteColor];
        _progressL.font = [UIFont systemFontOfSize:15];
        _progressL.textAlignment = NSTextAlignmentCenter;
        _progressL.layer.cornerRadius = 5.0;
        _progressL.hidden = YES;
        _progressL.backgroundColor = TImageMessageCell_Progress_Color;
        [_progressL.layer setMasksToBounds:YES];
        [self.container addSubview:_progressL];
        _progressL.frame = self.container.bounds;
        _progressL.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    }
    return self;
}

- (MSFlashImageMessageCellData *)flashImageData
{
    return (MSFlashImageMessageCellData *)self.flashImageData;
}

- (void)fillWithData:(MSFlashImageMessageCellData *)data
{
    [super fillWithData:data];
    NSInteger progress = data.flashElem.progress*100;
    self.progressL.text = [NSString stringWithFormat:@"%zd%%",progress];
    [self.progressL setHidden:!(progress > 0 && progress < 100)];
    BOOL isRead = data.flashElem.isSelf ? data.flashElem.from_see : data.flashElem.to_see;
    self.maskView.image = isRead ? [UIImage imageNamed:TUIKitResource(@"flashImg_sel")] : [UIImage imageNamed:TUIKitResource(@"flashImg_nor")];
    self.fireIcon.image = isRead ? [UIImage imageNamed:TUIKitResource(@"flashFire_sel")] : [UIImage imageNamed:TUIKitResource(@"flashFire_nor")];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    self.fireIcon.frame = CGRectMake(self.container.width * 0.5 - 25, self.container.height * 0.5 - 25, 50, 50);
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSFlashImageMessageCell.swift
//
//  MSFlashImageMessageCell.swift
//  MSIM-Swift
//
//  Created by benny wang on 2022/1/26.
//

import UIKit

open class MSFlashImageMessageCell: MSMessageCell {

    public var maskCoverView: UIImageView!
    
    public var fireIcon: UIImageView!
    
    public var progressL: UILabel!
    
    public var flashImageData: MSFlashImageMessageCellData?
    
    public override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        maskCoverView = UIImageView()
        maskCoverView.layer.cornerRadius = 5
        maskCoverView.layer.masksToBounds = true
        maskCoverView.contentMode = .scaleAspectFill
        container.addSubview(maskCoverView)
        maskCoverView.frame = container.bounds
        maskCoverView.autoresizingMask = [.flexibleWidth,.flexibleHeight]
        
        fireIcon = UIImageView()
        container.addSubview(fireIcon)
        
        progressL = UILabel()
        progressL.textColor = .white
        progressL.font = .systemFont(ofSize: 15)
        progressL.textAlignment = .center
        progressL.layer.cornerRadius = 5
        progressL.isHidden = true
        progressL.backgroundColor = MSMcros.TImageMessageCell_Progress_Color
        progressL.layer.masksToBounds = true
        container.addSubview(progressL)
        progressL.frame = container.bounds
        progressL.autoresizingMask = [.flexibleWidth,.flexibleHeight]
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public override func fillWithData(data: MSMessageCellData) {
        super.fillWithData(data: data)
        if let flashData = data as? MSFlashImageMessageCellData {
            self.flashImageData = flashData
            let progress = flashData.flashElem.progress * 100
            self.progressL.text = "\(progress)%"
            self.progressL.isHidden = !(progress > 0 && progress < 100)
            
            let isRead = flashData.flashElem.isSelf ? flashData.flashElem.from_see : flashData.flashElem.to_see
            self.maskCoverView.image = isRead ? UIImage.bf_imageNamed(name: "flashImg_sel") : UIImage.bf_imageNamed(name: "flashImg_nor")
            self.fireIcon.image = isRead ? UIImage.bf_imageNamed(name: "flashFire_sel") : UIImage.bf_imageNamed(name: "flashFire_nor")
        }
    }

    open override func layoutSubviews() {
        super.layoutSubviews()
        self.fireIcon.frame = CGRect(x: self.container.width * 0.5 - 25, y: self.container.height * 0.5 - 25, width: 50, height: 50)
    }
}
